# Artificial Neural Networks - Classification Example

Source: Adopted from https://www.pluralsight.com/guides/machine-learning-neural-networks-scikit-learn

### Step 1- Loading the Required Libraries and Modules

In [ ]:
# Import required libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor

# Import necessary modules
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from math import sqrt
from sklearn.metrics import r2_score

### Step 2 - Reading the Data and Performing Basic Data Checks

dataset source: https://www.kaggle.com/datasets/johndasilva/diabetes/versions/1?resource=download

In [ ]:
df = pd.read_csv('https://www.dropbox.com/s/ld8w9966o5o10ox/diabetes.csv?dl=1')
print(df.shape)
df.describe().transpose()

(2000, 9)


count       mean         std     min     25%  \
Pregnancies               2000.0    3.70350    3.306063   0.000   1.000   
Glucose                   2000.0  121.18250   32.068636   0.000  99.000   
BloodPressure             2000.0   69.14550   19.188315   0.000  63.500   
SkinThickness             2000.0   20.93500   16.103243   0.000   0.000   
Insulin                   2000.0   80.25400  111.180534   0.000   0.000   
BMI                       2000.0   32.19300    8.149901   0.000  27.375   
DiabetesPedigreeFunction  2000.0    0.47093    0.323553   0.078   0.244   
Age                       2000.0   33.09050   11.786423  21.000  24.000   
Outcome                   2000.0    0.34200    0.474498   0.000   0.000   

                              50%      75%     max  
Pregnancies                 3.000    6.000   17.00  
Glucose                   117.000  141.000  199.00  
BloodPressure              72.000   80.000  122.00  
SkinThickness              23.000   32.000  110.00  
Insulin                    40.000  130.000  744.00  
BMI                        32.300   36.800   80.60  
DiabetesPedigreeFunction    0.376    0.624    2.42  
Age                        29.000   40.000   81.00  
Outcome                     0.000    1.000    1.00

In [ ]:
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0               2      138             62             35        0  33.6   
1               0       84             82             31      125  38.2   
2               0      145              0              0        0  44.2   
3               0      135             68             42      250  42.3   
4               1      139             62             41      480  40.7   
...           ...      ...            ...            ...      ...   ...   
1995            2       75             64             24       55  29.7   
1996            8      179             72             42      130  32.7   
1997            6       85             78              0        0  31.2   
1998            0      129            110             46      130  67.1   
1999            2       81             72             15       76  30.1   

      DiabetesPedigreeFunction  Age  Outcome  
0                        0.127   47        1  
1                        0.233   23        0  
2                        0.630   31        1  
3                        0.365   24        1  
4                        0.536   21        0  
...                        ...  ...      ...  
1995                     0.370   33        0  
1996                     0.719   36        1  
1997                     0.382   42        0  
1998                     0.319   26        1  
1999                     0.547   25        0  

[2000 rows x 9 columns]

### Step 3 - Creating Arrays for the Features and the Response Variable

In [ ]:
target_column = ['Outcome']  # Create Object of Target Variable
predictors = list(set(list(df.columns))-set(target_column)) # List all features excewpt target variable
df[predictors] = df[predictors]/df[predictors].max() # Normalize predictors, i.e., scale b/w 0 and 1
df.describe().transpose() # see the data to check all features are between 0 and 1

count      mean       std       min       25%  \
Pregnancies               2000.0  0.217853  0.194474  0.000000  0.058824   
Glucose                   2000.0  0.608957  0.161149  0.000000  0.497487   
BloodPressure             2000.0  0.566766  0.157281  0.000000  0.520492   
SkinThickness             2000.0  0.190318  0.146393  0.000000  0.000000   
Insulin                   2000.0  0.107868  0.149436  0.000000  0.000000   
BMI                       2000.0  0.399417  0.101115  0.000000  0.339640   
DiabetesPedigreeFunction  2000.0  0.194599  0.133699  0.032231  0.100826   
Age                       2000.0  0.408525  0.145511  0.259259  0.296296   
Outcome                   2000.0  0.342000  0.474498  0.000000  0.000000   

                               50%       75%  max  
Pregnancies               0.176471  0.352941  1.0  
Glucose                   0.587940  0.708543  1.0  
BloodPressure             0.590164  0.655738  1.0  
SkinThickness             0.209091  0.290909  1.0  
Insulin                   0.053763  0.174731  1.0  
BMI                       0.400744  0.456576  1.0  
DiabetesPedigreeFunction  0.155372  0.257851  1.0  
Age                       0.358025  0.493827  1.0  
Outcome                   0.000000  1.000000  1.0

### Step 4 - Creating the Training and Test Datasets

In [ ]:
X = df[predictors].values # Create an array of features
y = df[target_column].values # create array of target

# split into test-train
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)
# see shape for test and train
print(X_train.shape); print(X_test.shape)

(1400, 8)
(600, 8)


### Step 5 - Building, Predicting, and Evaluating the Neural Network Model

we will build the neural network model using the scikit-learn library's estimator object, 'Multi-Layer Perceptron Classifier'

see https://keras.io/ for Guides

In [ ]:
import warnings
warnings.filterwarnings('ignore')

# Load the MLP Classifier
from sklearn.neural_network import MLPClassifier

#instantiates the model with 3 'hidden_layer_sizes' with 8 neurons each (for the 8 features)
# relu is activation function and
# adam is solver for weight optimization
# max_iter is number of iterations
mlp = MLPClassifier(hidden_layer_sizes=(8,8,8), activation='relu', solver='sgd', max_iter=200)

# Fit ANN model to training data
mlp.fit(X_train,y_train)

#Generate predictions on Training Set
predict_train = mlp.predict(X_train)
#Generate predictions on Testing Set
predict_test = mlp.predict(X_test)

#### Evaluate Performance on Training Data

In [ ]:
# Load Prediction Evaluation Functions
from sklearn.metrics import classification_report,confusion_matrix

# Print Confusion MAtrix for TRAINING
print("Confusion Matrix for Training Data \n", confusion_matrix(y_train,predict_train))
# Print Prediction Performance metrics for TRAINING
print("\n \n Prediction Performance for Training Data \n",classification_report(y_train,predict_train))

Confusion Matrix for Training Data 
 [[813 114]
 [182 291]]

 
 Prediction Performance for Training Data 
               precision    recall  f1-score   support

           0       0.82      0.88      0.85       927
           1       0.72      0.62      0.66       473

    accuracy                           0.79      1400
   macro avg       0.77      0.75      0.75      1400
weighted avg       0.78      0.79      0.78      1400



Support: classification population for  each class
i.e., in training data, there were 927 class 0 and 473 class 1

Weighted average implies performance is weighted by class population
Macro Avg is not weighted

#### Evaluate Performance on TESTING Data

In [ ]:
# Print Confusion MAtrix for TEST
print("Confusion Matrix for *TEST* Data \n", confusion_matrix(y_test,predict_test))
# Print Prediction Performance metrics for TRAINING
print("\n \n Prediction Performance for *TEST* Data \n",classification_report(y_test,predict_test))

Confusion Matrix for *TEST* Data 
 [[345  44]
 [ 64 147]]

 
 Prediction Performance for *TEST* Data 
               precision    recall  f1-score   support

           0       0.84      0.89      0.86       389
           1       0.77      0.70      0.73       211

    accuracy                           0.82       600
   macro avg       0.81      0.79      0.80       600
weighted avg       0.82      0.82      0.82       600



In [ ]:
# Load the MLP Classifier
from sklearn.neural_network import MLPClassifier

#instantiates the model with 3 'hidden_layer_sizes' with 8 neurons each (for the 8 features)
# relu is activation function and
# adam is solver for weight optimization
# max_iter is number of iterations
mlp = MLPClassifier(hidden_layer_sizes=(8,8,8), activation='relu', solver='adam', max_iter=1000)

# Fit ANN model to training data
mlp.fit(X_train,y_train)

#Generate predictions on Training Set
predict_train = mlp.predict(X_train)
#Generate predictions on Testing Set
predict_test = mlp.predict(X_test)

In [ ]:
# Print Confusion MAtrix for TRAINING
print("Confusion Matrix for Training Data \n", confusion_matrix(y_train,predict_train))
# Print Prediction Performance metrics for TRAINING
print("\n \n Prediction Performance for Training Data \n",classification_report(y_train,predict_train))

# Print Confusion MAtrix for TEST
print("Confusion Matrix for *TEST* Data \n", confusion_matrix(y_test,predict_test))
# Print Prediction Performance metrics for TRAINING
print("\n \n Prediction Performance for *TEST* Data \n",classification_report(y_test,predict_test))

Confusion Matrix for Training Data 
 [[825 102]
 [234 239]]

 
 Prediction Performance for Training Data 
               precision    recall  f1-score   support

           0       0.78      0.89      0.83       927
           1       0.70      0.51      0.59       473

    accuracy                           0.76      1400
   macro avg       0.74      0.70      0.71      1400
weighted avg       0.75      0.76      0.75      1400

Confusion Matrix for *TEST* Data 
 [[350  39]
 [ 82 129]]

 
 Prediction Performance for *TEST* Data 
               precision    recall  f1-score   support

           0       0.81      0.90      0.85       389
           1       0.77      0.61      0.68       211

    accuracy                           0.80       600
   macro avg       0.79      0.76      0.77       600
weighted avg       0.80      0.80      0.79       600

